### DISCUSSION TITLE
Can you make it so copy measure automatically starts with the selected area if there is one?

### DISCUSSION THREAD

**Discussion by Sy256** [helge17/tuxguitar#78] [General]
Currently, on Mac, Copy Measures seems to start with begin and end values at where the cursor is, even if a span of measures is selected. I'd like it to start from the beginning of the selection to the end of it. So if I have measures 14 to 18 selected, when I open the copy measure dialog it should show 14 to 18 as the default values, instead of 18 to 18 which it shows now. It was this way in 2.0 beta so hopefully its easy to pull that over. Thanks!

**Comment by guiv42** [comment]
No it's not possible, since there are no more 2 independent actions "copy" and "copy measure" accessible through the ui (see [this discussion](https://github.com/helge17/tuxguitar/discussions/29#discussion-5169828), for which I had explicitly [asked for feedback](https://github.com/helge17/tuxguitar/discussions/24#discussioncomment-5864927)).
When "copy" action is activated (through menu or shortcut):
- either nothing is selected, then "copy measure" dialog opens
- or something is selected, then it is copied directly, without opening any dialog

So, if you want to copy measures: select from first beat of first measure to last beat of last measure then copy, or don't select anything and copy to open dialog.

**Reply by Sy256** [reply]
But it's still there in the Measure menu on the latest version as a separate action that I have shortcutted differently than normal 'copy'. 

>So, if you want to copy measures: select from first beat of first measure to last beat of last measure then copy, or don't select anything and copy to open dialog.

The first way would just copy the selected track only though right? The strength of 'copy measure' is that it can copy all the tracks at once, which I need sometimes. And the paste menu allows for inserting and not overwriting. I'm sorry I think I missed the feedback thread on that and now that I'm in using this version for everything I can test it more so I notice these things.

I am still able to access the 'copy measure' action separately, is that because the mac version is different than the rest?

If it would be hard and require an overhaul to make it so highlighting a range starts the dialog at those measures, I will just get used to counting the measures manually to plug in, but I don't want to lose the separate action entirely.

Also is there any way I can help on coding? I know nearly nothing about code (dabbled in python and PHP once), but if I learned Java would I be able to help more directly/understand what is actually doable or is this really harder than anything I could learn relatively quickly?

**Comment by guiv42** [comment]
In fact I had not even noticed the copy measure dialog was still directly accessible...
So it should be feasible, I will have a look
And to answer one of your questions: no, the user interface is not supposed to be different on mac compared to Win or Linux

If you would like to contribute, I think there are other methods than diving into the code. Modifying someone else's code is not a comfortable activity (I rediscover it every day!), especially when there is no structured documentation about the code itself. Most probably a difficult way to start learning how to develop.

Still there are some other ways to contribute.
I think it could be profitable if you can spend some time testing more accurately the recently added feature (look at the commits history). Especially for features back-ported from 2.0beta that are not very reliable. I corrected some bugs on the fly, and even some times a complete re-design was needed.
Also you could test how 2.0beta fork applies all effects to a complete selection. I am currently starting the activity to port these features, it could be helpful if you can identify the existing bugs.
Another possible activity is to update the user documentation. helge17 has already done a great job in this activity, but user documentation is a kind of neverending story. This is plain html so you would not need to dive in the code. A good starting point could be all features related to the "click & drag" selection feature.

**Reply by helge17** [reply]
I'm still working on the documentation, but it will take a few more days before I have a consistent version I can check in.

**Reply by Sy256** [reply]
Hello I have some time to help now. 2 things, as I'm still new to github, what are commits and where do I find them? Are they like a update log for what's changed from version to version? Also how do I access the user documentation to add to it? Thanks!

**Reply by guiv42** [reply]
@Sy256:
1. at first you need to learn how git works and how to use it, you will find many tutorials on the net. Then you need to fork/clone this repo (see github documentation), and then build from source (see instructions [there](https://github.com/helge17/tuxguitar/blob/master/INSTALL.md)).
2. concerning user documentation, please read the message above: helge17 is currently working on it, so it's clearly not the right moment to do the same job concurrently.
3. can you please avoid to mix up too many different subjects in one thread? Just re-read the title of this thread: you asked for a feature, and it's been implemented (see my message below). So I think it would make sense to close this discussion now, and address non-related topics elsewhere (by the way I think you can also close #24). For example, I have created discussion #41 to discuss about everyone's contribution, with the objective to avoid concurrent activities. Just updated it with my current on-going activities.

**Reply by Sy256** [reply]
Ok thanks I will close this topic after clarifying these:

- I will learn how github works more
- I'll not touch the user documentation because helge17 is already working on it, so I can't do much to contribute I'll just standby and test for mac releases like I've been doing and try to learn java on the side for myself.
- I have one last question on the note repeating function thread before closing it
- The way I've been using this part of github "discussions" is for optional feature requests mostly (things that could make it better), and any bugs (things that break it and should be high priority) go in "issues" instead, is that right? I have one, not-that important bug in "issues" right now.
- I'll only make one feature request per thread from now on
- Should I be using the sourceforge feature requests section instead, is that where you're more likely to look? You guys have been great responding here, and I didn't know about the SF section so I will move my future requests over there if that's better organized for you. Or is it better to keep them here? And if I move future FR to the old FR on SF, what is "discussions" for?

**Reply by guiv42** [reply]
I agree, discussions seem more appropriate for feature request and issues for bug reports. And keeping 1 separate discussion for each feature request makes it easier to follow.
Please keep all of them only in this repository, **not** on SF (which is no more maintained by anyone)

**Reply by Sy256** [reply]
Got it thanks, closing this thread now!

**Comment by guiv42** [comment]
done (92493420b2fe87577a1e85ea9b19968231e6f17f)
